src/main/java/io/opentracing/contrib/concurrent/TracedExecutorService.java

```diff
@@ -59,101 +60,108 @@ public boolean awaitTermination(long l, TimeUnit timeUnit) throws InterruptedExc
 
   @Override
   public <T> Future<T> submit(Callable<T> callable) {
-    Scope scope = createScope("submit");
+    Span span = createSpan("submit");
     try {
-      return delegate.submit(tracer.activeSpan() == null ? callable :
-          new TracedCallable<T>(callable, tracer));
+      Span toActivate = span != null ? span : tracer.activeSpan();
+      return delegate.submit(toActivate == null ? callable :
+          new TracedCallable<T>(callable, tracer, toActivate));
     } finally {
-      if (scope != null) {
-        scope.close();
+      if (span != null) {
+        span.finish();
       }
     }
   }
 
   @Override
   public <T> Future<T> submit(Runnable runnable, T t) {
-    Scope scope = createScope("submit");
+    Span span = createSpan("submit");
     try {
-      return delegate.submit(tracer.activeSpan() == null ? runnable :
-          new TracedRunnable(runnable, tracer), t);
+      Span toActivate = span != null ? span : tracer.activeSpan();
+      return delegate.submit(toActivate == null ? runnable :
+          new TracedRunnable(runnable, tracer, toActivate), t);
     } finally {
-      if (scope != null) {
-        scope.close();
+      if (span != null) {
+        span.finish();
       }
     }
   }
 
   @Override
   public Future<?> submit(Runnable runnable) {
-    Scope scope = createScope("submit");
+    Span span = createSpan("submit");
     try {
-      return delegate.submit(tracer.activeSpan() == null ? runnable :
-          new TracedRunnable(runnable, tracer));
+      Span toActivate = span != null ? span : tracer.activeSpan();
+      return delegate.submit(toActivate == null ? runnable :
+          new TracedRunnable(runnable, tracer), toActivate);
```

Comment:
The `toActivate` should be in the `TracedRunnable` parameters I believe?

Reply:
good catch!